part_2/solution.py

```diff
@@ -0,0 +1,861 @@
+opt.output_nc = 1  # Number of output channels.
+opt.resize_or_crop = "none"  # Scaling and cropping of images at load time [resize_and_crop|crop|scale_width|scale_width_and_crop|none].
+opt.target = "nuclei"  # or "cyto" depending on your choice of target for virtual stain.
+
+# Load Training Set for input into model
+opt.isTrain = True 
+train_dataloader = CreateDataLoader(opt)
+dataset_train = train_dataloader.load_data()
+print(f"Total Training Images = {len(train_dataloader)}")
+
+# Load Val Set
+opt.phase = "val"
+val_dataloader = CreateDataLoader(opt)
+dataset_val = val_dataloader.load_data()
+print(f"Total Validation Images = {len(val_dataloader)}")
+opt.phase= "train"
+# %% [markdown]
+"""
+## Configure Pix2PixHD GAN and train to predict nuclei from phase.
+Having loaded the data into the model we can now train the Pix2PixHD GAN to predict nuclei from phase. We will use the following hyperparameters to train the model:
+
+"""
+# %%
+# Define the parameters for the Generator.
+opt.ngf = 64  # Number of filters in the generator.
+opt.n_downsample_global = 4  # Number of downsampling layers in the generator.
+opt.n_blocks_global = 9  # Number of residual blocks in the generator.
+opt.n_blocks_local = 3  # Number of residual blocks in the generator.
+opt.n_local_enhancers = 1  # Number of local enhancers in the generator.
+
+# Define the parameters for the Discriminators.
+opt.num_D = 3  # Number of discriminators.
+opt.n_layers_D = 3  # Number of layers in the discriminator.
+opt.ndf = 32  # Number of filters in the discriminator.
+
+# Define general training parameters.
+opt.gpu_ids = [0] # GPU ids to use.
+opt.norm = "instance"  # Normalization layer in the generator.
+opt.use_dropout = ""  # Use dropout in the generator (fixed at 0.2).
+opt.batchSize = 8  # Batch size.
+
+# Create a visualizer to perform image processing and visualization
+visualizer = Visualizer(opt)
+
+
+# Here will first start training a model from scrach however we can continue to train from a previously trained model by setting the following parameters.
+opt.continue_train = False
+if opt.continue_train:
+    iter_path = os.path.join(opt.checkpoints_dir, opt.name, "iter.txt")
+    try:
+        start_epoch, epoch_iter = np.loadtxt(iter_path, delimiter=",", dtype=int)
+    except:
+        start_epoch, epoch_iter = 1, 0
+    print("Resuming from epoch %d at iteration %d" % (start_epoch, epoch_iter))
+else:
+    start_epoch, epoch_iter = 1, 0
+    
+print('------------ Options -------------')
+for k, v in sorted(vars(opt).items()):
+    print('%s: %s' % (str(k), str(v)))
+print('-------------- End ----------------')
+
+# Set the number of epoch to be 1 for demonstration purposes
+opt.n_epochs = 2 # start from 1
+# Initialize the model
+phase2nuclei_model = create_model(opt)
+# Define Optimizers for G and D
+optimizer_G, optimizer_D = (
+    phase2nuclei_model.module.optimizer_G,
+    phase2nuclei_model.module.optimizer_D,
+)
+# %%
+train_model(
```

Comment:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[12], line 2
      1 # %%
----> 2 train_model(
      3     opt,
      4     phase2nuclei_model,
      5     visualizer,
      6     dataset_train,
      7     dataset_val,
      8     optimizer_G,
      9     optimizer_D,
     10     start_epoch,
     11     epoch_iter,
     12     writer,
     13 )

File ~/data/06_image_translation/part2/GAN_code/GANs_MI2I/pix2pixHD/train_dlmbl.py:209, in train(opt, model, visualizer, dataset_train, dataset_val, optimizer_G, optimizer_D, start_epoch, epoch_iter, writer)
    206 train_loss_D_fake, train_loss_D_real, train_loss_G_GAN, train_loss_G_Feat, train_loss_G_VGG, = train_epoch(opt, model, visualizer, dataset_train, optimizer_G, optimizer_D, total_steps, epoch, epoch_iter)
    208 [val_loss_D_fake, val_loss_D_real, val_loss_G_GAN, val_loss_G_Feat, val_loss_G_VGG], virtual_stain, fluorescence, brightfield = val_epoch(opt, model, dataset_val)
--> 209 visualizer.results_plot(brightfield,fluorescence,virtual_stain,['Phase Contrast', 'Fluorescence', 'Virtual Stain'],writer,epoch,rows=brightfield.shape[0])
    211 # Tensorboard Logging
    212 epoch_discriminator = {'fake_is_fake': train_loss_D_fake, 'real_is_real': train_loss_D_real}

File ~/data/06_image_translation/part2/GAN_code/GANs_MI2I/pix2pixHD/util/visualizer.py:190, in Visualizer.results_plot(self, input_x, target, predictions, titles, writer, epoch, rows)
    187 fig, axs = plt.subplots(input_x.shape[0], 3, figsize=(10, 30))
    189 # Set the titles for each column
--> 190 axs[0, 0].set_title(titles[0])
    191 axs[0, 1].set_title(titles[1])
    192 axs[0, 2].set_title(titles[2])

IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed
```

Reply:
I suggest you split this visualizer and show the stuff later.

Reply:
Hi Ed I have commited the fix for this to the https://github.com/Tonks684/GANs_MI2I repo (https://github.com/Tonks684/GANs_MI2I/commit/47c634474aba3b4c5ebb07eeb3272f3209a59ee0). It was due to the default batch being set to 1 (for speed of showing the training but the plot expected multiple images to display). The plot now works for batch size 1 and the default batch size has been set to 8 to show the students multiple images